File: packages/react-native-sdk/ios/TracksRecorder/TracksRecorderManager.swift
//
// Copyright © 2026 Stream.io Inc. All rights reserved.
//

import AVFoundation
import CoreMedia
import Foundation
import WebRTC

/// Orchestrator for the React Native track recorder. Owns the
/// `AVAssetWriter`, the recording lifecycle, the writer-start gate, and
/// the terminal-completion barrier. Delegates the encoder + sink + drain
/// work to `VideoPipeline` and `AudioPipeline` respectively (composed via
/// `PipelineHost`).
///
/// The `startRecording` completion is the *lifecycle* signal — it fires once at the terminal moment of the
/// recording (auto-stop timer expired, manual stop, or fatal error),
/// carrying the resulting file URL or an error. The `stopRecording`
/// completion is just a synchronization point — `void` — so callers can
/// `await stopTrackRecording(); await getStreamRecordings()` without
/// racing the disk flush.
@objc public final class TracksRecorderManager: NSObject, PipelineHost {

    @objc public static let shared = TracksRecorderManager()

    // MARK: - Configuration

    @objc public var recordingsDirectory: URL {
        let tmp = NSTemporaryDirectory()
        let url = URL(fileURLWithPath: tmp).appendingPathComponent("StreamRecordings", isDirectory: true)
        try? FileManager.default.createDirectory(
            at: url,
            withIntermediateDirectories: true,
            attributes: nil
        )
        return url
    }

    // MARK: - PipelineHost

    let queue = DispatchQueue(label: "io.stream.video.tracks-recorder")

    private(set) var assetWriter: AVAssetWriter?
    private(set) var isRecording = false

    // MARK: - State

    private var videoPipeline: VideoPipeline?
    private var audioPipeline: AudioPipeline?

    private var outputURL: URL?
    private var recordingCompletion: ((URL?, NSError?) -> Void)?
    private var isCompleted = false

    private var pendingPipelines = 0
    private var recordingStartHostTimeNs: UInt64?
    private var autoStopTimer: DispatchSourceTimer?

    // MARK: - Public API

    @objc public func startRecording(
        videoTrackId: String?,
        maxDurationMs: Int,
        targetWidth: Int,
        targetHeight: Int,
        webRTCModule: WebRTCModule,
        completion: @escaping (URL?, NSError?) -> Void
    ) {
        queue.async { [weak self] in
            guard let self = self else { return }

            if self.isRecording {
                completion(nil, makeRecorderError("recording_in_progress", code: 1))
                return
            }

            let videoTrack = videoTrackId
                .flatMap { webRTCModule.track(forId: $0) } as? RTCVideoTrack

            // Resolve the APM for the render-side audio tap before opening
            // any files — so a missing/incompatible APM fails fast. Audio
            // is captured through the APM render-pre delegate (post-mix
            // decoded audio); no per-track lookup is needed.
            let apmId = WebRTCModuleOptions.sharedInstance().audioProcessingModule
            guard let apm = apmId as? RTCDefaultAudioProcessingModule else {
                completion(nil, makeRecorderError("audio_processing_module_unavailable", code: 5))
                return
            }
            if apm.renderPreProcessingDelegate != nil {
                completion(nil, makeRecorderError("recording_blocked_by_audio_render_tap", code: 6))
                return
            }

            let dir = self.recordingsDirectory
            let timestamp = Int(Date().timeIntervalSince1970 * 1000)
            let outputURL = dir.appendingPathComponent("recording_\(timestamp).mp4")

            let writer: AVAssetWriter
            do {
                writer = try AVAssetWriter(url: outputURL, fileType: .mp4)
            } catch {
                completion(nil, error as NSError)
                return
            }

            self.assetWriter = writer
            self.outputURL = outputURL
            self.recordingCompletion = completion
            self.isCompleted = false
            self.recordingStartHostTimeNs = nil
            self.pendingPipelines = 0
            self.isRecording = true

            if let videoTrack = videoTrack {
                let pipeline = VideoPipeline(
                    host: self,
                    videoTrack: videoTrack,
                    targetWidth: Int32(targetWidth),
                    targetHeight: Int32(targetHeight)
                )
                self.videoPipeline = pipeline
                self.pendingPipelines += 1
                pipeline.start()
            }

            let audio = AudioPipeline(host: self, apm: apm)
            self.audioPipeline = audio
            self.pendingPipelines += 1
            audio.start()

            if maxDurationMs > 0 {
                let timer = DispatchSource.makeTimerSource(queue: self.queue)
                timer.schedule(deadline: .now() + .milliseconds(maxDurationMs))
                timer.setEventHandler { [weak self] in
                    self?.stopRecording { }
                }
                timer.resume()
                self.autoStopTimer = timer
            }

            NSLog("[TracksRecorder] recording started video=%@ audio=YES → %@",
                  videoTrack != nil ? "YES" : "NO",
                  outputURL.absoluteString)
        }
    }

    @objc public func stopRecording(completion: @escaping () -> Void) {
        queue.async { [weak self] in
            guard let self = self else {
                DispatchQueue.main.async { completion() }
                return
            }

            if !self.isRecording {
                DispatchQueue.main.async { completion() }
                return
            }

            self.autoStopTimer?.cancel()
            self.autoStopTimer = nil

            // Detach sinks/taps so no more frames arrive on the recorder path.
            self.videoPipeline?.detachSink()
            self.audioPipeline?.detachSink()

            let video = self.videoPipeline
            let audio = self.audioPipeline
            let writer = self.assetWriter
            let outputURL = self.outputURL

            self.isRecording = false

            // If the writer never started (no encoded sample ever made it
            // out), there's nothing to finalise.
            guard let assetWriter = writer, assetWriter.status == .writing else {
                NSLog("[TracksRecorder] stopRecording: writer not in .writing (status=%ld) — no file produced",
                      Int(writer?.status.rawValue ?? -1))
                video?.logSummary()
                audio?.logSummary()
                self.fireTerminalCompletion(url: nil, error: writer?.error as NSError?)
                self.cleanupAfterStop()
                DispatchQueue.main.async { completion() }
                return
            }

            video?.markInputAsFinished()
            audio?.markInputAsFinished()

            assetWriter.finishWriting { [weak self] in
                guard let self = self else {
                    DispatchQueue.main.async { completion() }
                    return
                }
                self.queue.async {
                    let resolved: URL? = (assetWriter.status == .completed) ? outputURL : nil
                    let writerError: NSError? = (assetWriter.status == .failed)
                        ? (assetWriter.error as NSError?)
                        : nil
                    video?.logSummary()
                    audio?.logSummary()
                    NSLog("[TracksRecorder] recording finalised → %@",
                          resolved?.absoluteString ?? "(no file produced)")
                    self.fireTerminalCompletion(url: resolved, error: writerError)
                    self.cleanupAfterStop()
                    DispatchQueue.main.async { completion() }
                }
            }
        }
    }

    @objc public func clearRecordingsDirectory(completion: @escaping (NSError?) -> Void) {
        queue.async { [weak self] in
            guard let self = self else { return }
            if self.isRecording {
                DispatchQueue.main.async {
                    completion(makeRecorderError("recording_in_progress", code: 1))
                }
                return
            }
            let fm = FileManager.default
            do {
                let contents = try fm.contentsOfDirectory(
                    at: self.recordingsDirectory,
                    includingPropertiesForKeys: nil,
                    options: []
                )
                for url in contents {
                    try? fm.removeItem(at: url)
                }
                DispatchQueue.main.async { completion(nil) }
            } catch {
                DispatchQueue.main.async { completion(error as NSError) }
            }
        }
    }

    @objc public func listRecordings() -> [URL] {
        let fm = FileManager.default
        guard let contents = try? fm.contentsOfDirectory(
            at: recordingsDirectory,
            includingPropertiesForKeys: [.contentModificationDateKey],
            options: [.skipsHiddenFiles]
        ) else {
            return []
        }
        return contents.sorted { lhs, rhs in
            let lDate = (try? lhs.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            let rDate = (try? rhs.resourceValues(forKeys: [.contentModificationDateKey]).contentModificationDate) ?? .distantPast
            return lDate > rDate
        }
    }

    // MARK: - PipelineHost

    func seedOriginNs(_ timestampNs: UInt64) -> UInt64 {
        if let existing = recordingStartHostTimeNs { return existing }
        recordingStartHostTimeNs = timestampNs
        return timestampNs
    }

    func onTrackAdded() {
        pendingPipelines = max(0, pendingPipelines - 1)
        maybeStartSession()
    }

    func onFatalError(_ error: NSError) {
        fireTerminalCompletion(url: nil, error: error)
        cleanupAfterFailure()
    }

    // MARK: - Internal helpers

    /// Starts the writer once every active pipeline has reported its input
    /// added. Calling `startWriting()` / `startSession(.zero)` before all
    /// inputs are present would leave un-added tracks orphaned; the gate
    /// is load-bearing.
    private func maybeStartSession() {
        guard let writer = assetWriter else { return }
        if pendingPipelines > 0 { return }
        if writer.status != .unknown { return }

        guard writer.startWriting() else {
            let err = (writer.error as NSError?) ?? makeRecorderError("start_failed", code: 3)
            fireTerminalCompletion(url: nil, error: err)
            cleanupAfterFailure()
            return
        }
        // The session origin is `.zero`; PTS values are computed relative
        // to `recordingStartHostTimeNs` before being passed to
        // `VTCompressionSessionEncodeFrame`, so the first encoded sample
        // already carries pts=0.
        writer.startSession(atSourceTime: .zero)
        NSLog("[TracksRecorder] writer.startWriting() + startSession(.zero)")
    }

    private func fireTerminalCompletion(url: URL?, error: NSError?) {
        guard !isCompleted else { return }
        isCompleted = true
        let cb = recordingCompletion
        recordingCompletion = nil
        DispatchQueue.main.async { cb?(url, error) }
    }

    private func cleanupAfterFailure() {
        videoPipeline?.detachSink()
        audioPipeline?.detachSink()
        autoStopTimer?.cancel()
        autoStopTimer = nil
        resetTransientState()
    }

    private func cleanupAfterStop() {
        resetTransientState()
    }

    /// Resets every transient field to its initial value. Single source of
    /// truth for "the manager is between recordings". Does NOT release
    /// native resources — the caller must have already torn down the
    /// pipelines and the writer.
    private func resetTransientState() {
        videoPipeline = nil
        audioPipeline = nil
        assetWriter = nil
        outputURL = nil
        recordingCompletion = nil
        isCompleted = false
        isRecording = false
        pendingPipelines = 0
        recordingStartHostTimeNs = nil
        autoStopTimer?.cancel()
        autoStopTimer = nil
    }
}

File: packages/react-native-sdk/android/src/main/java/com/streamvideo/reactnative/recorder/TracksRecorderManager.kt
package com.streamvideo.reactnative.recorder

import android.content.Context
import android.media.MediaMuxer
import android.os.Handler
import android.os.HandlerThread
import android.os.Looper
import android.util.Log
import com.oney.WebRTCModule.WebRTCModule
import java.io.File
import org.webrtc.VideoTrack

/**
 * Orchestrator for the React Native track recorder. Owns the [MediaMuxer], the recording lifecycle,
 * the muxer-start gate, and the terminal-completion barrier. Delegates the encoder + sink + drain
 * work to [VideoPipeline] and [AudioPipeline] respectively (composed via [PipelineHost]).
 *
 * The public surface is wrapped by `StreamVideoReactNativeModule.kt`'s `startTrackRecording` /
 * `stopTrackRecording` / `clearStreamRecordings` / `getStreamRecordings` methods, which in turn are
 * the bridge contract used by the JS `useLoopbackRecording` hook. Knows nothing about loopback or
 * any specific recording use case — it's a generic encode-and-mux orchestrator.
 *
 * Both video and audio are optional. Audio is always requested by the current bridge contract;
 * video is requested whenever the caller passes a `videoTrackId` that resolves to a [VideoTrack].
 * The [MediaMuxer] only starts once **all** active pipelines have reported their track (gated by
 * [maybeStartMuxer] on [pendingPipelines]).
 *
 * Threading: a dedicated [HandlerThread] serialises every state mutation (start, stop, encoder
 * feed, muxer writes) so the rest of the file is lock-free. Pipelines accept buffers on WebRTC
 * delivery threads and post to this handler before touching any state.
 *
 * Completion semantics: `startRecording` is the **lifecycle promise** — it fires once at the
 * recording's terminal moment with the produced file (or an error). `stopRecording` is a void sync
 * point that resolves after native finalisation, so callers can `await stopTrackRecording(); await
 * getStreamRecordings()` without racing the disk flush. Same shape as iOS.
 */
class TracksRecorderManager private constructor() : PipelineHost {

    companion object {
        @JvmField val shared = TracksRecorderManager()

        private const val TAG = "TracksRecorder"
        private const val RECORDINGS_DIR_NAME = "StreamRecordings"
    }

    private val thread = HandlerThread("io.stream.video.tracks-recorder").apply { start() }
    private val timerHandler = Handler(Looper.getMainLooper())

    override val handler = Handler(thread.looper)
    override var muxer: MediaMuxer? = null
        private set
    override var muxerStarted = false
        private set

    private var videoPipeline: VideoPipeline? = null
    private var audioPipeline: AudioPipeline? = null

    private var outputFile: File? = null

    private var recordingCompletion: ((File?, Throwable?) -> Unit)? = null
    private var isCompleted = false
    private var isRecording = false
   
    private var pendingPipelines = 0
    private var recordingStartHostTimeNs: Long? = null
    private var autoStopRunnable: Runnable? = null

    fun recordingsDirectory(context: Context): File {
        val dir = File(context.cacheDir, RECORDINGS_DIR_NAME)
        if (!dir.exists()) dir.mkdirs()
        return dir
    }

    fun startRecording(
            context: Context,
            webRTCModule: WebRTCModule,
            videoTrackId: String?,
            maxDurationMs: Long,
            targetWidth: Int = 0,
            targetHeight: Int = 0,
            completion: (File?, Throwable?) -> Unit,
    ) {
        handler.post {
            if (isRecording) {
                completion(null, RecordingError("recording_in_progress"))
                return@post
            }

            val resolvedVideoTrack =
                    videoTrackId?.let { webRTCModule.getTrackById(it) } as? VideoTrack

            val dir = recordingsDirectory(context)
            val outFile = File(dir, "recording_${System.currentTimeMillis()}.mp4")
            val muxerInstance: MediaMuxer =
                    try {
                        MediaMuxer(
                                outFile.absolutePath,
                                MediaMuxer.OutputFormat.MUXER_OUTPUT_MPEG_4
                        )
                    } catch (t: Throwable) {
                        completion(null, t)
                        return@post
                    }

            resetTransientState()

            this.muxer = muxerInstance
            this.outputFile = outFile
            this.recordingCompletion = completion
            this.isRecording = true

            if (resolvedVideoTrack != null) {
                val pipeline = VideoPipeline(
                        host = this,
                        videoTrack = resolvedVideoTrack,
                        targetWidth = targetWidth,
                        targetHeight = targetHeight,
                )
                videoPipeline = pipeline
                pendingPipelines++
                pipeline.start()
            }

            val audio = AudioPipeline(host = this, webRTCModule = webRTCModule)
            audioPipeline = audio
            pendingPipelines++
            audio.start()

            if (maxDurationMs > 0) {
                val runnable = Runnable { stopRecording { /* fire-and-forget */} }
                autoStopRunnable = runnable
                timerHandler.postDelayed(runnable, maxDurationMs)
            }

            Log.i(
                    TAG,
                    "recording started video=${resolvedVideoTrack != null} audio=true → ${outFile.absolutePath}",
            )
        }
    }

    fun stopRecording(completion: () -> Unit) {
        // Detach sinks synchronously off the recorder handler so no
        // new buffers can be enqueued while the backlog drains. The
        // audio pipeline also restores speaker volume here so the
        // mute lifts immediately.
        videoPipeline?.detachSink()
        audioPipeline?.detachSink()
        autoStopRunnable?.let { timerHandler.removeCallbacks(it) }
        autoStopRunnable = null

        handler.post {
            if (!isRecording) {
                completion()
                return@post
            }

            val video = videoPipeline
            val audio = audioPipeline
            val muxerInstance = muxer
            val resolved = outputFile

            if (!muxerStarted || muxerInstance == null) {
                Log.w(TAG, "stopRecording: muxer never started — discarding empty recording")
                video?.logSummary()
                audio?.logSummary()
                video?.stopAndRelease()
                audio?.stopAndRelease()
                try {
                    muxerInstance?.release()
                } catch (_: Throwable) {}
                // Best-effort: delete the empty file so getStreamRecordings()
                // doesn't surface an unplayable 0-byte mp4.
                resolved?.delete()
                fireTerminalCompletion(null, null)
                cleanupAfterStop()
                completion()
                return@post
            }

            // Skip the EOS drain when EOS can't be queued — waiting on
            // a marker that will never arrive would hang the handler.
            if (video != null) {
                try {
                    val queued = video.signalEndOfStream(muxerInstance)
                    if (queued) video.drainAfterEoS(muxerInstance)
                } catch (t: Throwable) {
                    Log.e(TAG, "stopRecording: video drain failed", t)
                }
            }
            if (audio != null) {
                try {
                    val queued = audio.signalEndOfStream(muxerInstance)
                    if (queued) audio.drainAfterEoS(muxerInstance)
                } catch (t: Throwable) {
                    Log.e(TAG, "stopRecording: audio drain failed", t)
                }
            }

            video?.stopAndRelease()
            audio?.stopAndRelease()

            var finalResolved: File? = resolved
            try {
                muxerInstance.stop()
            } catch (t: Throwable) {
                Log.e(TAG, "muxer.stop() threw — likely no usable samples", t)
                finalResolved = null
                resolved?.delete()
            }
            try {
                muxerInstance.release()
            } catch (t: Throwable) {
                Log.w(TAG, "muxer.release() threw", t)
            }

            video?.logSummary()
            audio?.logSummary()
            Log.i(
                    TAG,
                    "recording finalised → ${finalResolved?.absolutePath ?: "(no file produced)"}",
            )

            fireTerminalCompletion(finalResolved, null)
            cleanupAfterStop()
            completion()
        }
    }

    fun clearRecordingsDirectory(context: Context, completion: (Throwable?) -> Unit) {
        handler.post {
            try {
                val dir = recordingsDirectory(context)
                dir.listFiles()?.forEach { it.deleteRecursively() }
                completion(null)
            } catch (t: Throwable) {
                completion(t)
            }
        }
    }

    fun listRecordings(context: Context): List<File> {
        val dir = File(context.cacheDir, RECORDINGS_DIR_NAME)
        if (!dir.isDirectory) return emptyList()
        return dir.listFiles()?.sortedByDescending { it.lastModified() } ?: emptyList()
    }

    override fun seedOriginNs(timestampNs: Long): Long {
        val existing = recordingStartHostTimeNs
        if (existing != null) return existing
        recordingStartHostTimeNs = timestampNs
        return timestampNs
    }

    override fun onTrackAdded() {
        pendingPipelines = (pendingPipelines - 1).coerceAtLeast(0)
        muxer?.let { maybeStartMuxer(it) }
    }

    override fun onFatalError(error: Throwable) {
        fireTerminalCompletion(null, error)
        cleanupAfterFailure()
    }

    /**
     * Starts the muxer once every active pipeline has added its track. Calling `start()` before all
     * `addTrack` calls makes subsequent `addTrack` throw "Muxer is not initialized", so the gate is
     * load-bearing.
     */
    private fun maybeStartMuxer(muxerInstance: MediaMuxer) {
        if (muxerStarted) return
        if (pendingPipelines > 0) return
        
        try {
            muxerInstance.start()
            muxerStarted = true
        } catch (t: Throwable) {
            Log.e(TAG, "muxer.start() threw", t)
            fireTerminalCompletion(null, t)
            cleanupAfterFailure()
        }
    }

    private fun fireTerminalCompletion(file: File?, error: Throwable?) {
        if (isCompleted) return
        isCompleted = true

        val cb = recordingCompletion
        recordingCompletion = null
        cb?.invoke(file, error)
    }

    /**
     * Resets every transient field to its initial value. Single source of truth for "the manager is
     * between recordings". Does NOT release native resources — the caller must stop/release
     * encoders and the muxer before invoking this.
     */
    private fun resetTransientState() {
        muxer = null
        muxerStarted = false
        videoPipeline = null
        audioPipeline = null
        outputFile = null
        recordingCompletion = null
        isCompleted = false
        isRecording = false
        pendingPipelines = 0
        recordingStartHostTimeNs = null
        autoStopRunnable?.let { timerHandler.removeCallbacks(it) }
        autoStopRunnable = null
    }

    private fun cleanupAfterFailure() {
        videoPipeline?.detachSink()
        audioPipeline?.detachSink()
        videoPipeline?.stopAndRelease()
        audioPipeline?.stopAndRelease()
        try {
            muxer?.release()
        } catch (t: Throwable) {
            Log.w(TAG, "failed to release muxer", t)
        }
        resetTransientState()
    }

    private fun cleanupAfterStop() {
        resetTransientState()
    }
}
